src/Storages/Hive/StorageHive.cpp

```diff
@@ -0,0 +1,667 @@
+    auto hive_table_metadata = hive_metastore_client->getTableMetadata(hive_database, hive_table);
+
+    /// List files under partition directory in HDFS
+    auto list_paths = [hive_table_metadata, &fs](const String & path) { return hive_table_metadata->getFilesByLocation(fs, path); };
+
+    std::vector<Apache::Hadoop::Hive::Partition> partitions = hive_table_metadata->getPartitions();
+    /// Hive files to read
+    HiveFiles hive_files;
+    /// Mutext to protect hive_files, which maybe appended in multiple threads
+    std::mutex hive_files_mutex;
+
+    auto append_hive_files = [&](const HiveMetastoreClient::FileInfo & hfile, const FieldVector & fields)
+    {
+        LOG_TRACE(log, "Append hive file {}", hfile.path);
+        String filename = getBaseName(hfile.path);
+
+        /// Skip temporary files starts with '.'
+        if (filename.find('.') == 0)
+            return;
+
+        auto file = createHiveFile(
+            format_name,
+            fields,
+            hdfs_namenode_url,
+            hfile.path,
+            hfile.last_modify_time,
+            hfile.size,
+            hivefile_name_types,
+            storage_settings,
+            context_);
+
+        /// Load file level minmax index and apply
+        const KeyCondition hivefile_key_condition(query_info, getContext(), hivefile_name_types.getNames(), hivefile_minmax_idx_expr);
+        if (file->hasMinMaxIndex())
+        {
+            file->loadMinMaxIndex();
+            if (!hivefile_key_condition.checkInHyperrectangle(file->getMinMaxIndex()->hyperrectangle, hivefile_name_types.getTypes())
+                     .can_be_true)
+            {
+                LOG_TRACE(log, "Skip hive file {} by index {}", file->getPath(), file->describeMinMaxIndex(file->getMinMaxIndex()));
+                return;
+            }
+        }
+
+        /// Load sub-file level minmax index and apply
+        std::set<int> skip_splits;
+        if (file->hasSubMinMaxIndex())
+        {
+            file->loadSubMinMaxIndex();
+            const auto & sub_minmax_idxes = file->getSubMinMaxIndexes();
+            for (size_t i = 0; i < sub_minmax_idxes.size(); ++i)
+            {
+                if (!hivefile_key_condition.checkInHyperrectangle(sub_minmax_idxes[i]->hyperrectangle, hivefile_name_types.getTypes())
+                         .can_be_true)
+                {
+                    LOG_TRACE(log, "Skip split {} of hive file {}", i, file->getPath());
+                    skip_splits.insert(i);
+                }
+            }
+            file->setSkipSplits(skip_splits);
+        }
+
+        {
+            std::lock_guard lock{hive_files_mutex};
+            hive_files.push_back(file);
+        }
+    };
+
+    ThreadPool pool{num_streams};
+    if (!partitions.empty())
+    {
+        const auto partition_names = partition_name_types.getNames();
+        const auto partition_types = partition_name_types.getTypes();
+
+        for (const auto & partition : partitions)
```

Comment:
I think it is better to move this part to a separate method. `read` method now is too big, so better to split.

Reply:
Ok, I will try.

Reply:
done.